src/roman_datamodels/datamodels/_asdf.py

```diff
@@ -0,0 +1,181 @@
+from __future__ import annotations
+
+import sys
+from collections.abc import Callable, Generator
+from contextlib import contextmanager
+from os import PathLike
+from pathlib import Path
+from typing import Any, ParamSpec, TypeVar, cast
+
+from asdf import AsdfFile
+from asdf.search import AsdfSearchResult
+
+from roman_datamodels.stnode import DNode, TaggedObjectNode
+from roman_datamodels.validate import nuke_validation
+
+__all__ = ["AsdfFileMixin"]
+
+_T = TypeVar("_T")
+_P = ParamSpec("_P")
+
+
+@contextmanager
+def _temporary_update_filename(datamodel: AsdfFileMixin[_T], filename: PathLike[str] | str) -> Generator[None, None, None]:
+    """
+    Context manager to temporarily update the filename of a datamodel so that it
+    can be saved with that new file name without changing the current model's filename
+    """
+    from roman_datamodels.nodes import Filename
+
+    dm: DNode[DNode[_T]] = cast(DNode[DNode[_T]], datamodel)
+
+    if "meta" in cast(DNode[_T], dm.fields) and "filename" in dm.meta.fields:
+        old_filename = dm.meta.filename
+        dm.meta.filename = cast(_T, Filename(filename))
+
+        yield
+        dm.meta.filename = old_filename
+        return
+
+    yield
+    return
+
+
+class AsdfFileMixin(DNode[DNode[_T] | _T]):
+    _is_copy: bool = False
+    _asdf: AsdfFile | None = None
+
+    @classmethod
+    def node_type(cls) -> type:
+        """
+        Get the top-level node type for this model
+        -> this is assumed to be the last class in the multiple inheritance
+        """
+        return cls.__bases__[-1]
+
+    def construct_node(self, data: dict[str, DNode[_T] | _T]) -> TaggedObjectNode:
+        """
+        Construct a new node (matching the node for this model) from the given data
+        """
+        # MyPy can't correctly infer the typing here
+        return self.node_type()(data, _array_shape=self._array_shape_)  # type: ignore[no-any-return]
+
+    def construct_model(self, data: dict[str, DNode[_T] | _T] | DNode[DNode[_T] | _T] | None) -> AsdfFileMixin[_T]:
+        """
+        Construct a new model from the given data
+        """
+        return type(self)(data, _array_shape=self._array_shape_)
+
+    def _init_asdf_file(self) -> None:
+        """
+        Initialize the ASDF file
+        """
+        with nuke_validation():
+            # ASDF has not implemented type hints so MyPy will complain about this
+            # until they do.
+            af = AsdfFile()  # type: ignore[no-untyped-call]
+
+            af["roman"] = self.construct_node(self._data)
+            # ASDF has not implemented type hints so MyPy will complain about this
+            # until they do.
+            af.validate()  # type: ignore[no-untyped-call]
```

Comment:
I left this in because it is the existing behavior; however, this will cause the creation of all the fields that need to be filled in for a valid file. This destroys the laziness in many cases and I don't think its removal will make any significant differences.